### DISCUSSION TITLE
Need a scraper to get information from .pyd files

### DISCUSSION THREAD

**Discussion by daltairwalter** [microsoft/pylance-release#2706] [Enhancement]
I notice that pyd files don't provide any information through pylance.  This is not surprising to me because to get information they would need to load the pyd files which would be a security risk.  I saw a previous request that suggested making a stub.  This seems like something that would be automated and it didn't suggest what utility I should run to create this stub.  Is there a python utility I can run to generate stubs for pyd files?

**Comment by jakebailey** [comment]
This advice is most useful for library authors who are making the pyd files as they know the interfaces that should be used. For consumers who can't control that, the advice may not be as helpful, but you can still add a stub for it if you know the interfaces you use. (For comparison, I've done this a number of times to add types for JavaScript libraries I want to use in TypeScript; the process is similar.)

To autogenerate, you'd need a python script that inspects things and makes a skeleton, yes. There's one that MPLS uses, but it doesn't output in a stub format. There's also a tool called stubgen which may work for compiled modules to produce a skeleton to manually edit, but I didn't have much luck when I tried it last.

There's a scraper I wrote based on the one in MPLS (that actually outputs conforming stubs), but it's not included in Pylance because we haven't added the ability to run it, and we wanted to get that whole process (when to run it, how to ask, where to store things, etc...) spec'd about before offering it to everyone.

We already have the ability to generate stubs for modules regular Python modules, but at the moment we don't do this for compiled code. The intent is to resolve this, but it's a nontrivial issue.

**Comment by daltairwalter** [comment]
I am looking for something that I could run as part of my pyd build process.  This is primarily for the pyd's that I create.  Is the scraper script that you have on github somewhere?

**Comment by erictraut** [comment]
A "scraper" cannot generate high-quality stubs because it lacks type information about function parameters and return types. High-quality stubs should also incorporate generics or Protocols where appropriate and function overloads. They should export only the symbols that are part of the officially-supported public interface. And in the case of pyd-based implementations, the stubs can also incorporate doc strings that can be displayed in smart editors.

A stub generator can provide a "first draft" of your type stubs, but most of the work involved in creating high-quality stubs is manual work. The good news is that once it's done, it's relatively easy and cheap to maintain stubs.

**Comment by jakebailey** [comment]
It is not, no, but I think that the scraper wouldn't be a good choice if you're the author of the compiled modules. The scraper is more of a last-resort case where "we have a pyd and we need to do something with it even if the output is inaccurate". The scraper is good at discovering the existence of things, but often not good at knowing the signatures without attempting to parse docstrings as libraries often are not compiling in their annotations. Now that we have Pylance which integrates type checking, inaccuracies in the stubs are more impactful (whereas MPLS didn't care if you mismatched things).

What do you use to make the `pyd`s? Is your project / build somewhere we could use as an example?

I think most people are preparing `.pyi` files which sit in the repo next to the uncompiled code, then shipping them next to their `pyd` files. If you're using cython, it'd be nice if it had the ability to output stubs (as it definitely knows the types).

**Comment by erictraut** [comment]
Here are some additional [tips about stub authoring](https://github.com/microsoft/pyright/blob/master/docs/type-stubs.md) that you might find useful.

**Comment by daltairwalter** [comment]
We have tried a number of strategies for making pyd's and still have SWIG, manually created, and pybind11 modules.  I honestly can't even imagine trying to maintain stub bindings - the maintenance burdens, inaccuracies, etc., not to mention the initial investment just isn't worth doing in the first place.  I am looking for something for something automated that provides whatever is possible.  If there is a way to annotate my C++ code that flows into the pyd and then can be scraped to make this better, I may be interested - but a manual solution is way too expensive.

**Comment by ldorigo** [comment]
If extracting types from .pyx files is undoable, could you consider to at least scrape  docstrings so that when libraries provide stubs corresponding to the .pyx files, the correct docstrings are displayed in vscode? It currently works with the older language server, but not pylance.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by daltairwalter** [comment]
For my purposes, I have found a good solution to this problem in pybind11-stubgen which automatically makes stub pyi files from pyd files.  Some of this may also be possible in mypy stubgen.  As I stated in my original post, this would be a security concern if pylance just loaded these files in.  It would still be helpful if pylance could guide users to one of these scraper tools or include one in the installation to make things more seamless.

**Reply by JMPZ11** [reply]
I have seen multiple references to loading pyx/pyd files as a "security" concern. How is it any more of a security risk than loading a normal python file? Normal python filed are also code that could contain malicious bits and pieces; pylance not reading .pyx or .pyd files in the name of "security" is a lame excuse

**Reply by erictraut** [reply]
Pylance loads Python code (text) files but doesn't execute them. Their content is analyzed statically. The same cannot be done with an arbitrary binary file like a "pyx" or "pyd".

**Comment by JMPZ11** [comment]
I may be missing something - in the project I'm working with, the .pyx and
.pyd files are also source code

**Reply by erictraut** [reply]
You can open the "pyd" in a text editor? I think you'll find that it's a binary dynamically-linked library. The only way to reason about it is to run it. But that comes with security risks because the code could do anything (delete files, introduce malware, upload private data, etc.).

The "pyx" is text, but it's not Python code. It would require a static analyzer different from pylance to analyze its contents. There are tools that know how to read and process ".pyx" files, and they could be modified to output Python-compatible type stubs. Pylance and other Python static analysis tools would then be able to consume it.

**Reply by JMPZ11** [reply]
My mistake they are .pxd and .pyx files; not .pyd.
Examples:
https://github.com/commaai/opendbc/blob/master/can/common.pxd
https://github.com/commaai/opendbc/blob/master/can/parser_pyx.pyx

This is Python 3.8, it uses a pipenv virtualenv and VSCode is using the virtualenv. I rarely encounter red squiggles, but frequently encounter unbound types coming from Cython code

pylance is not parsing these file for me - is it supposed to? If yes, I'll keep digging to figure out why it isn't working.... If no - I guess I go rant in the appropriate place :)

My apologies. I spent several hours banging my head on the gently-documented cython (and the abomination that c++ is for me...) syntax to create a pyi, and was quite frustrated...

**Reply by erictraut** [reply]
Pylance parses ".py" and ".pyi" files, both of which use Python syntax and semantics. It recognizes the presence of ".pyd" and ".so" files for purposes of import resolution, but it doesn't attempt to parse their contents because they are binaries, not text.

Pylance doesn't know how to parse other programming languages. If a library is compiled from another programming language (C++, C, Fortran, etc.), its interface should be described with a ".pyi" stub file. If a stub file is present, pylance will use it for completion suggestions, semantic highlight, and type checking.

If you don't have a ".pyi" file for your binary library, you may be able to generate a basic stub file using a tool like (stubgen)[https://mypy.readthedocs.io/en/stable/stubgen.html]. The resulting stub file will be incomplete (e.g. missing parameter type and return type annotations), but you may find that it's better than nothing. You could also use this generated stub as a "first draft" which you could then manually modify.

**Reply by JMPZ11** [reply]
I have more questions, but they are not in the scope of this issue, so I shall move along. Thanks for you help!

**Comment by Cirn09** [comment]
I found out that Vim's famous autocompletion plugin - `jedi` - implements the ability to get API information from pyd files.
After testing and auditing, `jedi` will only load `pyd` that is already **installed**.
I think user install a module means that the user trusts that module because it can be run at any time.

Implementing a restricted `pyd` scraper is necessary because there are a large number of Python packages distributed as binaries that do not provide PYI files. Even when pyi files are provided, they usually don't include a docstring (even `stubgen` doesn't do automatic docstring generation (https://github.com/python/mypy/pull/13284)).
Since loading pyd directly may cause performance problems, I recommend providing a switch option for this feature.


<details><summary> Implementation details for <code>jedi</code></summary>
<p>

Remove the "current directory" from the Python path:
https://github.com/davidhalter/jedi/blob/77cfefc1cc25d41b10018c3457ad91e445c2622a/jedi/api/project.py#L187-L189
Import `pyd`
https://github.com/davidhalter/jedi/blob/77cfefc1cc25d41b10018c3457ad91e445c2622a/jedi/inference/compiled/access.py#L118-L120
`pyd` module blacklist:
https://github.com/davidhalter/jedi/blob/77cfefc1cc25d41b10018c3457ad91e445c2622a/jedi/inference/compiled/__init__.py#L62-L66

</p>
</details>

**Reply by Cirn09** [reply]
any update?

**Comment by SheepDomination** [comment]
I'm new to stubbing; and I tool would like to know how do I stub a `*pyd` file as so that I can use it with autocomplete ?
